src/leo_nlm_lock_handler.erl

```diff
@@ -0,0 +1,366 @@
+    gen_server:call(?MODULE, {unlock, FileHandler, Owner, Start, End}).
+
+%% ---------------------------------------------------------------------
+%% GEN_SERVER CALLBACKS
+%% ---------------------------------------------------------------------
+init(_Args) ->
+    ets:new(?NLM_LOCK_ETS, [set, named_table, private]),
+    {ok, void}.
+
+start_link(_Args)->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+handle_call({test, FileHandler, Lock}, _From, State) ->
+    case test_lock(FileHandler, Lock) of
+        {ok, _} ->
+            {reply, ok, State};
+        {{error, Lock}, _} ->
+            {reply, {error, Lock}, State}
+    end;
+handle_call({lock, FileHandler, Lock}, _From, State) ->
+    case test_lock(FileHandler, Lock) of
+        {ok, ExistingLocks} ->
+            ets:insert(?NLM_LOCK_ETS, {FileHandler, [Lock | ExistingLocks]}),
+            {reply, ok, State};
+        {{error, Lock}, _} ->
+            {reply, {error, Lock}, State}
+    end;
+handle_call({unlock, FileHandler, Owner, Start, End}, _From, State) ->
+    Ret = case ets:lookup(?NLM_LOCK_ETS, FileHandler) of
+              [{FileHandler, ExistingLocks}] ->
+                  NewLocks = lists:foldl(
+                               fun(CurLock, Acc) ->
+                                       case CurLock#lock_record.owner of
+                                           Owner ->
+                                               case modify_lock(CurLock, Start, End) of
+                                                   [] ->
+                                                       Acc;
+                                                   Modified ->
+                                                       lists:append(Acc, Modified)
+                                               end;
+                                           _ ->
+                                               [CurLock | Acc]
+                                       end
+                               end, [], ExistingLocks),
+                  ets:insert(?NLM_LOCK_ETS, {FileHandler, NewLocks});
+              _ ->
+                  ok
+          end,
+    {reply, Ret, State}.
+
+handle_cast(Req, S) ->
+    ?debug("handle_cast", "req:~p", [Req]),
+    {noreply, S}.
+
+handle_info(Req, S) ->
+    ?debug("handle_info", "req:~p", [Req]),
+    {noreply, S}.
+
+terminate(_Reason, _S) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% ---------------------------------------------------------------------
+%% INNER FUNCTIONS
+%% ---------------------------------------------------------------------
+-spec(test_lock(binary(), #lock_record{}) ->
+        {ok, list(#lock_record{})} | {{error, #lock_record{}}, list(#lock_record{})}).
+test_lock(FileHandler, Lock) ->
+    case ets:lookup(?NLM_LOCK_ETS, FileHandler) of
+        [] ->
+            {ok, []};
+        [{FileHandler, ExistingLocks}] ->
+            {check_lock_ranges(Lock, ExistingLocks), ExistingLocks}
+    end.
+
```

Comment:
Since those inner functions (check_lock_ranges, modify_lock) seem to be shared with other impls,
Putting into one common module would be good.